pkg/path/extract.go

```diff
@@ -0,0 +1,305 @@
+			readWritePaths = append(readWritePaths, readWritePathChildren...)
+		} else if dirEntry.IsList() {
+			itemPath = formatName(dirEntry, true, parentPath, subpathPrefix)
+			if dirEntry.Config == yang.TSFalse || parentState == yang.TSFalse {
+				subpathPfx := subpathPrefix
+				if parentState == yang.TSFalse {
+					subpathPfx = itemPath[len(parentPath):]
+				}
+				readOnlyPathsChildren, _, err := extractPaths(dirEntry, yang.TSFalse, parentPath, subpathPfx)
+				if err != nil {
+					return nil, nil, err
+				}
+				readOnlyPaths = append(readOnlyPaths, readOnlyPathsChildren...)
+				continue
+			}
+			readOnlyPathsChildren, readWritePathsChildren, err := extractPaths(dirEntry, dirEntry.Config, itemPath, "")
+			if err != nil {
+				return nil, nil, err
+			}
+
+			readOnlyPaths = append(readOnlyPaths, readOnlyPathsChildren...)
+			readWritePaths = append(readWritePaths, readWritePathsChildren...)
+			// Need to copy the index of the list across to the RO list too
+			//roIdxName := k[:strings.LastIndex(k, "/")]
+			//roIdxSubPath := k[strings.LastIndex(k, "/"):]
+			//indices, _ := ExtractIndexNames(itemPath[strings.LastIndex(itemPath, "/"):])
+			//isIdxAttr := false
+			//for _, idx := range indices {
+			//	if roIdxSubPath == fmt.Sprintf("/%s", idx) {
+			//		isIdxAttr = true
+			//	}
+			//}
+			//if roIdxName == itemPath && isIdxAttr {
+			//	roIdx := ReadOnlyAttrib{
+			//		ValueType:   v.ValueType,
+			//		Description: v.Description,
+			//		Units:       v.Units,
+			//	}
+			//	readOnlyPaths[roIdxName] = make(map[string]ReadOnlyAttrib)
+			//	readOnlyPaths[roIdxName][roIdxSubPath] = roIdx
+			//}
+
+		} else if dirEntry.IsChoice() || dirEntry.IsCase() {
+			// Recurse down through Choice and Case
+			readOnlyPathsTemp, readWritePathsTemp, err := extractPaths(dirEntry, dirEntry.Config, parentPath, "")
+			if err != nil {
+				return nil, nil, err
+			}
+			readOnlyPaths = append(readOnlyPaths, readOnlyPathsTemp...)
+			readWritePaths = append(readWritePaths, readWritePathsTemp...)
+		} else {
+			log.Warnf("Unexpected type of leaf for %s %v", itemPath, dirEntry)
+		}
+	}
+
+	return readOnlyPaths, readWritePaths, nil
+}
+
+func formatName(dirEntry *yang.Entry, isList bool, parentPath string, subpathPrefix string) string {
+	parentAndSubPath := parentPath
+	if subpathPrefix != "/" {
+		parentAndSubPath = fmt.Sprintf("%s%s", parentPath, subpathPrefix)
+	}
+
+	var name string
+	if isList {
+		//have to ensure index order is consistent where there's more than one
+		keyParts := strings.Split(dirEntry.Key, " ")
+		sort.Slice(keyParts, func(i, j int) bool {
+			return keyParts[i] < keyParts[j]
+		})
+		name = fmt.Sprintf("%s/%s", parentAndSubPath, dirEntry.Name)
+		for _, k := range keyParts {
+			name += fmt.Sprintf("[%s=*]", k)
+		}
+	} else {
+		name = fmt.Sprintf("%s/%s", parentAndSubPath, dirEntry.Name)
+	}
+
+	return name
+}
+
+func toValueType(entry *yang.YangType, isLeafList bool) (configapi.ValueType, []uint64, error) {
```

Comment:
I do have pretty much the same method here: https://github.com/onosproject/config-models/blob/master/pkg/gnmi-client-gen/utils.go#L90

We should try to consolidate at a certain point

Reply:
This was written a long time ago here https://github.com/onosproject/onos-config/blob/be9fe145adaece61bb521b4d3817b1108214b471/pkg/utils/path/path.go#L370